include/rtio.h

```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_RTIO_H_
+#define ZEPHYR_INCLUDE_RTIO_H_
+/**
+ * @brief RTIO Interface
+ * @defgroup rtio_interface RTIO Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
+ * provide functionality to safely encode or decode blocks of memory.
+ */
+
+/**
+ * @brief Chunk of memory with ownership and byte format versioning
+ */
+struct rtio_block {
+	/* Mem block from k_mem_pool */
+	struct k_mem_block_id mem_block_id;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write began
+	 */
+	u32_t begin_tstamp;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write was complete
+	 */
+	u32_t end_tstamp;
+
+	/** Byte layout designator given by each driver to specify
+	 * the layout of the contained buffer
+	 */
+	u32_t layout;
+
+	/** Length of data in this buffer. */
+	u16_t len;
+
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
+	/** Start of the buffer data .*/
+	u8_t *_data;
+};
+
+/**
+ * @brief Initialize an rtio_block
+ */
+__syscall void rtio_block_init(struct rtio_block *block,
+				   struct k_mem_block_id mem_block_id,
+			       u8_t *data, u32_t size);
+
+static inline void z_impl_rtio_block_init(struct rtio_block *block,
+				   struct k_mem_block_id mem_block_id,
+				   u8_t *data, u32_t size)
+{
+	block->mem_block_id = mem_block_id;
+	block->layout = 0;
+	block->begin_tstamp = 0;
+	block->end_tstamp = 0;
+	block->len = 0;
+	block->size = size;
+	block->_data = data;
+}
+
+/**
+ * @brief Begin a block read or write
+ */
+__syscall void rtio_block_begin(struct rtio_block *block);
+
+static inline void z_impl_rtio_block_begin(struct rtio_block *block)
+{
+	block->len = 0;
+	block->begin_tstamp = k_cycle_get_32();
+}
+
+/**
+ * @brief End a block read or write.
+ */
+__syscall void rtio_block_end(struct rtio_block *block);
+
+static inline void z_impl_rtio_block_end(struct rtio_block *block)
+{
+	block->end_tstamp = k_cycle_get_32();
+}
+
+
+/**
+ * @brief Unused number of bytes in the buffer
+ */
+__syscall u16_t rtio_block_available(struct rtio_block *block);
```

Comment:
Why do you need a syscall here? do you plan to put the block into "kernel space"?

Reply:
I added __syscalls on suggestion from @andrewboie but I don't think I thought about the implications fully, I've added a more detailed comment about the syscall interface below and question I have about it

Reply:
You should only add __syscall on functions that need the privileges to write or read memory objects that don't belong to the calling user thread. If you have to do so, you must check the provided pointers. See other syscall handlers for example.

Reply:
I've removed this as a syscall, I believe only the rtio_api backed functions (rtio_configure, rtio_trigger) need to be syscalls and thats exactly what I've changed